File: lib/utilis/geo.dart
import 'dart:math' as math;

/// Straight-line miles between two WGS84 points.
double milesBetween(
  double lat1,
  double lng1,
  double lat2,
  double lng2,
) {
  const earthRadiusMiles = 3958.7613;
  final dLat = _toRad(lat2 - lat1);
  final dLng = _toRad(lng2 - lng1);
  final a = _sin2(dLat / 2) +
      math.cos(_toRad(lat1)) * math.cos(_toRad(lat2)) * _sin2(dLng / 2);
  return 2 * earthRadiusMiles * math.asin(math.min(1, math.sqrt(a)));
}

/// City-speed ETA label matching Google-style strings (`5 mins`).
String formatDriverEta(double miles) {
  const avgCitySpeedMph = 20;
  final minutes = ((miles / avgCitySpeedMph) * 60).round().clamp(1, 24 * 60);
  if (minutes < 60) {
    return minutes == 1 ? '1 min' : '$minutes mins';
  }
  final hours = minutes ~/ 60;
  final remainder = minutes % 60;
  final hourLabel = hours == 1 ? '1 hour' : '$hours hours';
  if (remainder == 0) return hourLabel;
  final minLabel = remainder == 1 ? '1 min' : '$remainder mins';
  return '$hourLabel $minLabel';
}

double _toRad(double deg) => deg * math.pi / 180;

double _sin2(double x) {
  final s = math.sin(x);
  return s * s;
}

File: functions/src/shared/geo.ts
/** Straight-line miles between two WGS84 points. */
export function milesBetween(
  lat1: number,
  lng1: number,
  lat2: number,
  lng2: number,
): number {
  const toRad = (deg: number) => (deg * Math.PI) / 180;
  const earthRadiusMiles = 3958.7613;
  const dLat = toRad(lat2 - lat1);
  const dLng = toRad(lng2 - lng1);
  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRad(lat1)) *
      Math.cos(toRad(lat2)) *
      Math.sin(dLng / 2) *
      Math.sin(dLng / 2);
  return 2 * earthRadiusMiles * Math.asin(Math.min(1, Math.sqrt(a)));
}

/** City-speed ETA label matching Google-style strings (`5 mins`). */
export function formatDriverEta(miles: number): string {
  const avgCitySpeedMph = 20;
  const minutes = Math.max(1, Math.round((miles / avgCitySpeedMph) * 60));
  if (minutes < 60) {
    return minutes === 1 ? "1 min" : `${minutes} mins`;
  }
  const hours = Math.floor(minutes / 60);
  const remainder = minutes % 60;
  const hourLabel = hours === 1 ? "1 hour" : `${hours} hours`;
  if (remainder === 0) return hourLabel;
  const minLabel = remainder === 1 ? "1 min" : `${remainder} mins`;
  return `${hourLabel} ${minLabel}`;
}
